Improvements to maven-protoc-plugin

sergei-ivanov opened the pull request:
Hello David,

I am currently in a process of migrating our project from an elaborate ant+ivy build to maven. Unfortunately protobufs have a pretty subprime support for maven, but this time I decided to take a hit, allocate some spare time and contribute to your plugin (which seems to be the most alive of various clones of the original plugin). I have an intention to bring it to production grade and hopefully host it in the central repo eventually.

Here we go. I am landing a pretty big aggregated commit, containing over a week's work. The major changes/additions are:
1. Support for protobuf toolchains -- allows you to remove hardwired or property-based protocExecutable from project poms. Toolchains also support versioning for tools, thus you can define multiple versions of protoc executable in the toolchains.xml and then select the desired one by version or version range.
2. Mojo documentation and additional logging.
3. Site documentation (still need to create more content to cover various use cases)
4. Integration tests (these require toolchains to be set up)

The plugin operation is fully backwards compatible with the previous version. Maven3 is required to build the plugin (this is because otherwise site plugin does not work properly). The plugin still can be used in Maven2 projects (maven version 2.0.9 and above).

Please give it a careful review and I hope you'll accept this pull request.

Kind regards,
Sergei

sergei-ivanov committed:
added toolchain dependency to pom

sergei-ivanov committed:
support for protobuf tool chain through plugin extensions

sergei-ivanov committed:
additional logging before proto compilation

sergei-ivanov committed:
moved integration tests into a separate profile

sergei-ivanov committed:
added integration test

sergei-ivanov committed:
a template for plugin web site

sergei-ivanov committed:
restored the default behaviour of attempting to locate protoc in the $PATH if all else fails; updated integration tests

sergei-ivanov committed:
plugin version update, documentation update

sergei-ivanov committed:
corrected plugin name in logs

sergei-ivanov committed:
marked goals as thread safe (for maven3 concurrent builds) and updated mojo documentation

sergei-ivanov committed:
plexus component descriptor is automatically generated from annotations; documentation plugins included into pom

sergei-ivanov committed:
allowed protocExecutable to be specified on command line

sergei-ivanov committed:
maven site documentation

sergei-ivanov committed:
updated plugin management, reports, release profile; enforced maven build environment; updated test dependencies

sergei-ivanov committed:
upgraded google-collections to 1.0

sergei-ivanov committed:
added options to skip/force mojo execution

sergei-ivanov committed:
FAQ updated

sergei-ivanov committed:
mojo documentation updated

sergei-ivanov committed:
updated invoker settings to search only one level down (allows to set up reactor integration tests)

sergei-ivanov committed:
updated default goals for integration test

sergei-ivanov committed:
integration tests for proto imports across independent projects (jar import)

sergei-ivanov committed:
integration tests for proto test imports across independent projects (test-jar import)

sergei-ivanov committed:
integration tests for proto test imports in reactor projects

sergei-ivanov committed:
added a .gitignore file

mfriedenhagen commented:
Hello there,
just for the record:

I just pulled `git pull git://github.com/sergei-ivanov/maven-protoc-plugin.git master` 
and it compiled cleanly, all tests run successfullly, see http://huschteguzzel.de/hudson/job/maven-protoc-plugin/2/console

Regards
Mirko

mfriedenhagen commented:
I now reran using the run-its profile and there seem to be test failures in the integration test, see:
http://huschteguzzel.de/hudson/job/maven-protoc-plugin/3/console
Hm, that seems to be because I did not set up toolchains. However I only get protoc version 2.2 for my ubuntu version.

mfriedenhagen commented:
I now reran `mvn clean install -Prun-its` on my Mac, using:

```
[mirko@borg ~]$ cat ~/.m2/toolchains.xml 
<toolchains>
    <toolchain>
       <type>protobuf</type>
       <provides>
           <version>2.4.1</version>
       </provides>
       <configuration>
          <protocExecutable>/usr/local/bin/protoc</protocExecutable>
       </configuration>
    </toolchain>
</toolchains>
```

and the tests ran fine.

sergei-ivanov commented:
That's right, you will need to set up a protobuf toolchain in order to run integration tests.

Here's a link to the documentation on your own Jenkins instance:
http://huschteguzzel.de/hudson/job/maven-protoc-plugin/site/examples/protobuf-toolchain.html

There are two options for you:
1. Build a 2.4.x protoc for your target platform and set up a toolchain to point to it.
2. Set up another toolchain for your existing protoc 2.2 (the advantage of toolchains is that you can have any number of versioned protoc instances specified in your toolchains.xml). Downgrade _both_ protoc version in the enforcer _and_ protobuf dependency to [2.2, 2.3) in the tests. Protobuf java library and protoc compiler must have matching major versions, otherwise the generated code may not compile.

sergei-ivanov commented:
I am glad that you managed to run the integration tests in the end!

mfriedenhagen commented:
Hey there, I installed 2.4.1 manually and now the tests are running in Jenkins as well: http://huschteguzzel.de/hudson/job/maven-protoc-plugin/4/console
Hopefully dtrott will pick up your changes.

sergei-ivanov committed:
Enabled test protobuf definitions to import production definitions from the same module.
This was a bug spotted by binkley, who also checked in a bug fix on a separate branch.
Here is a slightly re-engineered fix and a couple of integration tests to validate the results.

sergei-ivanov committed:
Moved debug logging inside protoc compiler wrapper

sergei-ivanov committed:
Some minor clean-up

sergei-ivanov committed:
Renamed Protoc.compile() to execute()

sergei-ivanov committed:
Added configuration for C++ and Python output

sergei-ivanov committed:
Updated URLs and paths in the POM to reflect the current fork

sergei-ivanov committed:
Updated site deployment configuration; pegged plugin versions; upgraded report plugins and test dependencies

sergei-ivanov committed:
Temporarily disabled JXR report to reduce the size of generated site content

pnyheim commented:
FYI:
I have created a [pull-request](https://github.com/dtrott/maven-protoc-plugin/pull/16) that fixes the issue mentioned in TEST-5, so if this one is applied, the fail-never line can be commented out.

sergei-ivanov committed:
Fixed importing protobuf definitions from other projects or modules when namespaces are used

sergei-ivanov committed:
Merged pending changes: inclusion of main output directory into the list of test protobuf imports.

sergei-ivanov committed:
Updated site documentation and meta-data ahead of release

sergei-ivanov committed:
[maven-release-plugin] prepare release maven-protoc-plugin-0.2.0

sergei-ivanov committed:
[maven-release-plugin] prepare for next development iteration

sergei-ivanov committed:
Updated repo URL

Commit:
Added support for writing file descriptor sets.

sergei-ivanov committed:
Merge pull request #3 from rwicke/master

Support for file descriptor set output

sergei-ivanov committed:
Reverted some pom changes after the latest pull, added entries to .gitignore

sergei-ivanov committed:
Changed default paths for output directories and binary protobuf descriptor file

sergei-ivanov committed:
Upgraded maven version to 2.2.1, upgraded plugins, updated changes description

sergei-ivanov committed:
Added support for M2E 1.1+ integration

sergei-ivanov committed:
Deleted obsolete script

sergei-ivanov committed:
Merged work in progress (support for C++ and Python output)

sergei-ivanov committed:
Added logging for the new protoc parameters

sergei-ivanov committed:
A quick check for empty input to avoid unnecessary actions

sergei-ivanov committed:
Converted plugin metadata from Plexus javadoc annotations to Java 5 Mojo annotations

sergei-ivanov committed:
Normalized line endings

rwicke committed:
added support for protoc plugins

rwicke committed:
use --plugin arguments to locate plugin executables instead of modifying PATH

rwicke committed:
Added support for protoc plugins that are written in Java and deployed as Maven artifacts.

rwicke committed:
fixed typo

rwicke committed:
Added support for plugin arguments and JVM options

rwicke committed:
javaHome is now located via jdk toolchain if available

rwicke committed:
small logging changes

rwicke committed:
Bumped version number to 0.3.5-SNAPSHOT

rwicke committed:
Added documentation for descriptor set output and protoc plugins.

rwicke committed:
changed version back to 0.3.0-SNAPSHOT

sergei-ivanov committed:
Merge pull request #9 from rwicke/plugin-support

Added support for protoc plugins

sergei-ivanov committed:
Added rwicke to committers list, updated changes description

sergei-ivanov committed:
Minor code clean-up and documentation

sergei-ivanov committed:
- demoted some plugin logging to debug level
- extracted method for javaHome detection
- protected access level for all methods of AbstractProtocMojo
- private access level for fields of AbstractProtocMojo that are not used in subclasses

sergei-ivanov committed:
Used an simpler alternative utility method

sergei-ivanov committed:
Updated maven site documentation

sergei-ivanov committed:
Upgraded dependencies and maven plugins

sergei-ivanov committed:
[maven-release-plugin] prepare release maven-protoc-plugin-0.3.0

sergei-ivanov committed:
[maven-release-plugin] prepare for next development iteration

sergei-ivanov committed:
Fixed #10: Downgraded plexus-utils from 3.0.10 to 2.1 to make plugin compatible with M2E

rwicke committed:
added support for protoc plugin execution on Windows 64bit JVMs

rwicke committed:
Improved detection of Windows JVM data model.
Added documentation for winJvmDataModel option.

rwicke committed:
suppress WinRun4J error popup
added debug logging to protoc plugin assembler

sergei-ivanov committed:
Merge pull request #12 from rwicke/master

Added support for 64-bit JVMs in Java-based protoc plugins on Windows

rwicke committed:
added integration test for custom protoc plugin support

sergei-ivanov committed:
Merge pull request #13 from rwicke/protoc-plugin-it

Added integration test for generated protoc plugin

sergei-ivanov committed:
Updated list of changes ahead of release, reformatted all code

sergei-ivanov committed:
[maven-release-plugin] prepare release maven-protoc-plugin-0.3.1

sergei-ivanov committed:
[maven-release-plugin] prepare for next development iteration

sergei-ivanov committed:
Fixed #15: removed Aether dependency and made the plugin compatible with Maven 3.1.x

sergei-ivanov committed:
[maven-release-plugin] prepare release maven-protoc-plugin-0.3.2

sergei-ivanov committed:
[maven-release-plugin] prepare for next development iteration

sergei-ivanov committed:
Implemented #5: Add support for generating non-java output (C++ and Python)

sergei-ivanov committed:
#5: Renamed goal 'testCompile' to 'test-compile' for naming consistency with the new goals

sergei-ivanov committed:
#5: Added new goals to M2E lifecycle mapping

sergei-ivanov committed:
#5: Update site documentation

abatalev committed:
#19 Make clearing out of the output directory optional

sergei-ivanov committed:
Merge pull request #24 from abatalev/master

#19 Make clearing out of the output directory optional

sergei-ivanov committed:
Rolled future version forward to 0.4.0

sergei-ivanov committed:
#19 follow-up to pull request merge: documentation updated and missing if statement added

sergei-ivanov committed:
#26 made java6 the minimum required version

sergei-ivanov committed:
#27 Made Maven 3.0 a minimum requirement, adjusted dependencies, updated plugin tools configuration

sergei-ivanov committed:
#26 Target bytecode version is 1.6

sergei-ivanov committed:
#25 Fixed incompatibility with Maven 3.2.5 and above

sergei-ivanov committed:
#27 Updated plugins to the latest versions

sergei-ivanov committed:
#28 removed obsolete taglet handlers (all mojo metadata is now annotation-driven)

sergei-ivanov committed:
#28 fixed doclint issues

sergei-ivanov committed:
#26 Target bytecode version in integration tests is 1.6

sergei-ivanov committed:
#27 Updated plugins to the latest versions

sergei-ivanov committed:
#27 Updated enforcer rules and dependency logging

sergei-ivanov committed:
#28 fixed doclint issues, tweaked the reporting and javadoc settings

sergei-ivanov committed:
#26 Added java toolchain

sergei-ivanov committed:
#26 #27 Cleaned up IT POMs, used common parent with java toolchain and maven 3 requirements

sergei-ivanov committed:
Fixed ITs for compatibility with protoc 2.6

sergei-ivanov committed:
#26 Removed the last traces of Java 1.5

sergei-ivanov committed:
#26 #27 Cleaned up the code after Java 6 and Maven 3 became a requirement

sergei-ivanov committed:
Updated changes report

sergei-ivanov committed:
#27 Replaced deprecated API usage with the new Maven3 API

sergei-ivanov committed:
#27 Updated build plugins

sergei-ivanov committed:
Change log and change attributions

sergei-ivanov committed:
[maven-release-plugin] prepare release maven-protoc-plugin-0.4.0

sergei-ivanov committed:
[maven-release-plugin] prepare for next development iteration

sergei-ivanov committed:
#27 Updated build plugins and dependencies (the latest invoker plugin supports building under Maven 3.3.1 and later)

sergei-ivanov committed:
Converted readme file into markdown, added Bintray repo configuration

sergei-ivanov committed:
#30 Made some injected parameters available to subclasses, fixed the output of warnings from protoc

sergei-ivanov committed:
#30 Added new goals to support custom native protoc plugins

sergei-ivanov committed:
#10 Updated the FAQ entry

sergei-ivanov committed:
Minor logging change

sergei-ivanov committed:
#32 Ensure that the resources are attached in accordance with inclusion/exclusion patterns

sergei-ivanov committed:
#32 Added new parameter 'attachProtoSources'

sergei-ivanov committed:
#32 Added integration tests

sergei-ivanov committed:
#30 Run wrapped java plugins by explicitly specifying the plugin executable to protoc

sergei-ivanov committed:
#21 Add an option to attach descriptor set to the build

sergei-ivanov committed:
#30 #33 Implemented resolution of native protoc and plugin binaries as artifacts from Maven repo

sergei-ivanov committed:
Improved and updated site documentation

sergei-ivanov committed:
Used GitHub site plugin

sergei-ivanov committed:
Configured Bintray as deployment repository

sergei-ivanov committed:
[maven-release-plugin] prepare release maven-protoc-plugin-0.4.1

sergei-ivanov committed:
[maven-release-plugin] prepare for next development iteration

sergei-ivanov committed:
Fixed automatic site deployment during release

sergei-ivanov committed:
#34 Used LinkedHashSet everywhere for faster iteration with a consistent predictable order

sergei-ivanov committed:
Upgraded maven plugins to the latest version and updated the release settings

sergei-ivanov committed:
[maven-release-plugin] prepare release maven-protoc-plugin-0.4.2

sergei-ivanov committed:
[maven-release-plugin] prepare for next development iteration

sergei-ivanov committed:
Upgraded to the latest version of os-maven-plugin that has been fixed to support Maven 3.0.x

kedzie committed:
Added javanano mojos

-added integration test

sergei-ivanov committed:
Merge pull request #38 from kedzie/master

Added javanano mojo

sergei-ivanov committed:
#30 Added missing entries to Eclipse lifecycle definition

sergei-ivanov committed:
Removed redundant package declaration

sergei-ivanov committed:
#38 Added javanano to the list of built-in plugins in the documentation

sergei-ivanov committed:
#38 Allowed to specify command-line options for JavaNano generator

sergei-ivanov committed:
#38 Fixed the integration test

sergei-ivanov committed:
#38 Documentation updates

sergei-ivanov committed:
#38 Added an integration test for test-compile-javanano mojo

sergei-ivanov committed:
#30 Added an integration test for custom protoc plugins

sergei-ivanov committed:
Preparing for the release

sergei-ivanov committed:
[maven-release-plugin] prepare release maven-protoc-plugin-0.4.3

sergei-ivanov committed:
[maven-release-plugin] prepare for next development iteration

atorstling committed:
support for including source info in descriptor set

atorstling committed:
review changes

sergei-ivanov committed:
Merge pull request #39 from atorstling/include_source_info

support for including source info in descriptor set

sergei-ivanov committed:
Preparation for the release, documentation update

sergei-ivanov committed:
[maven-release-plugin] prepare release maven-protoc-plugin-0.4.4

sergei-ivanov committed:
[maven-release-plugin] prepare for next development iteration

sergei-ivanov committed:
Removed the sources and added relocation information

sergei-ivanov committed:
[maven-release-plugin] prepare release maven-protoc-plugin-0.5.0

sergei-ivanov committed:
[maven-release-plugin] prepare for next development iteration

sergei-ivanov committed:
Removed the remaining sources

sergei-ivanov commented:
Development moved to https://github.com/xolstice/protobuf-maven-plugin
